### PR TITLE
Docs: Add public API reference section

### LINKED ISSUE #22006: Categorize SymPy Modules
To avoid having a big TOC, I propose categorizing the modules in the SymPy Modules Reference section of the documentation.

cc @asmeurer @oscarbenjamin

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
This PR adds docs for the ``Public API `` section.

#### References to other Issues or PRs

Fixes #22006
See also #22007

It is based on master 

#### Brief description of what is fixed or changed

The docs in the directories ``modules`` have been re-categorized, see [Categorize SymPy Modules](https://github.com/sympy/sympy/issues/22006)

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by nanjekyejoannah** [comment]
@asmeurer @oscarbenjamin I think I will need help to understand how the doc tests work before I fix this PR for whatever is failing here. The docs build fine AFAICT, but the tests are failing.

**Comment by asmeurer** [comment]
You will need to update the paths here https://github.com/sympy/sympy/blob/master/sympy/testing/runtests.py#L654-L655

File: doc/src/modules/index.rst [hunk 1]
```diff
@@ -12,56 +12,15 @@ and read the tests for it, that will tell you everything you need to know.
 Most of the things are already documented though in this document, that is
 automatically generated using SymPy's docstrings.
 
-Click the  "modules" (:ref:`modindex`) link in the top right corner to easily
+Click the modules" (:ref:`modindex`) link in the top right corner to easily
 access any SymPy module, or use the list below:
```
Comment (asmeurer) [inline doc/src/modules/index.rst:16]: Let's delete all this text and write something better.

File: doc/src/modules/private/index.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+.. _private_reference:
+
+Private API reference
+=====================
```
Comment (asmeurer) [inline doc/src/modules/private/index.rst:4]: Let's not include this as an empty section.

**Comment by asmeurer** [comment]
This is going to create a lot of merge conflicts the longer it stays open. Let's merge this as soon as we can and fix any issues in later PRs.

**Comment by nanjekyejoannah** [comment]
I added the redirects but also remembered that we needed to change the reorganization. I created a separate branch here; https://github.com/nanjekyejoannah/sympy/tree/publicref_edit because am not very convinced the TOC should be this big yet that same page will have details for the private API.

**Comment by asmeurer** [comment]
I think instead of having separate index pages for all the sections, we should have one "SymPy API Reference" landing page with a "Public API" header and then links to "basics", "logic", etc. with descriptions for each one.

**Comment by asmeurer** [comment]
I think this is ready to go but I just want to double check if @nanjekyejoannah or @oscarbenjamin have any changes first.

**Comment by oscarbenjamin** [comment]
Looks good to me. I'll merge this now. Thanks @nanjekyejoannah!

We need some feedback before considering where to take this next.

**Pull request merged**